tests: net: socket: tcp: Add test to validate connect timeout

ssharks opened the pull request:
When there is no response from the server, a client side connect
should return a ETIMEOUT. This tests breaks the connection and validates
this behavior.

This tests validates an element that is already correctly implemented.

ssharks committed:
tests: net: socket: tcp: Add test to validate connect timeout

When there is no response from the server, a client side connect
should return a ETIMEOUT. This tests breaks the connection and validates
this behavior.

Signed-off-by: Sjors Hettinga <s.a.hettinga@gmail.com>

ssharks commented on tests/net/socket/tcp/src/main.c:
```diff
@@ -723,6 +723,42 @@ void test_open_close_immediately(void)
+
+	prepare_sock_tcp_v4(CONFIG_NET_CONFIG_MY_IPV4_ADDR, ANY_PORT,
+			    &c_sock, &c_saddr);
+	prepare_sock_tcp_v4(CONFIG_NET_CONFIG_MY_IPV4_ADDR, SERVER_PORT,
+			    &s_sock, &s_saddr);
```
You are right I have to remove the lines 739, 740

rlubos commented on tests/net/socket/tcp/src/main.c:
```diff
@@ -723,6 +723,45 @@ void test_open_close_immediately(void)
 	k_sleep(TCP_TEARDOWN_TIMEOUT);
 }
 
+void test_open_obstructed(void)
```
I think we could use more descriptive name, simply `test_connect_timeout()` is more specific IMO.

rlubos commented on tests/net/socket/tcp/src/main.c:
```diff
@@ -723,6 +723,45 @@ void test_open_close_immediately(void)
+
+	loopback_set_packet_drop_ratio(1.0f);
+
+	zassert_not_equal(connect(c_sock, (struct sockaddr *)&s_saddr,
```
I think it'd be better to check if the return code is equal to -1, which is an expected value in case of a failure.

rlubos commented on tests/net/socket/tcp/src/main.c:
```diff
@@ -723,6 +723,45 @@ void test_open_close_immediately(void)
+				  sizeof(s_saddr)),
+			  0, "connect succeed");
+
+	zassert_equal(errno, ETIMEDOUT,
+		"connect should be timed out, got %i", errno);
```
Please fix the alignment

ssharks replied:
Done

rlubos commented on tests/net/socket/tcp/src/main.c:
```diff
@@ -723,6 +723,45 @@ void test_open_close_immediately(void)
+
+	test_close(c_sock);
+
+	k_sleep(TCP_TEARDOWN_TIMEOUT);
```
I don't think we need teardown delay in this case, not connection has been established.

ssharks replied:
Removed

rlubos reviewed (commented):
I'd consider reducing `CONFIG_NET_SOCKETS_CONNECT_TIMEOUT` time to reduce the test time.

ssharks commented:
Pushed on top of an older commit, since there is a problem in the current main with a conflicting definition of can_set_bitrate

ssharks committed:
tests: net: socket: tcp: Make test_open_close_immediately more robust

The close function on a TCP socket can return before the complete
socket has been closed, as there might be packets still in flight.
Modify the test to wait briefly (less then the retransmit time), before
counting the number of still open net_contexts. This makes the test
outcome not dependent on the scheduling order of the different tasks.

Secondly the test actually checks the number of open contexts to be zero
so there is no need to wait for any open contexts to still close.

Signed-off-by: Sjors Hettinga <s.a.hettinga@gmail.com>

ssharks committed:
tests: net: socket: tcp: Make test_vX_msg_waitall more robust

The close function on a TCP socket can return before the complete
socket has been closed, as there might be packets still in flight.
Add a wait at the end of the test so we are sure all sockets have
been closed before the next test starts.

Signed-off-by: Sjors Hettinga <s.a.hettinga@gmail.com>

ssharks commented:
@rlubos It took a few iterations to fix seemingly unrelated bugs in the test suite. Could you have a look at the current state?

ssharks commented:
@jukkar Could you have a look at this PR?

jukkar commented on tests/net/socket/tcp/src/main.c:
```diff
@@ -709,18 +709,29 @@ void test_open_close_immediately(void)
+	/* Allow for the close communication to finish,
+	 * this makes the test success, no longer scheduling dependent
+	 */
+	k_sleep(K_MSEC(CONFIG_NET_TCP_INIT_RETRANSMISSION_TIMEOUT/2));
```
I would suggest that you set the `CONFIG_NET_TCP_INIT_RETRANSMISSION_TIMEOUT` in the test config file just in case the default value changes. This way we could avoid long timeouts etc hard to debug issues later (not likely but still on option).

ssharks replied:
Totally agree, fixed

jukkar reviewed (commented):
Looks good, one small suggestion to change.

rlubos commented on tests/net/socket/tcp/src/main.c:
```diff
@@ -709,18 +709,66 @@ void test_open_close_immediately(void)
+	/* Allow for the close communication to finish,
+	 * this makes the test success, no longer scheduling dependent
+	 */
+	k_sleep(K_MSEC(CONFIG_NET_TCP_INIT_RETRANSMISSION_TIMEOUT/2));
```
I'm surprised compliance does not report that:
```suggestion
	k_sleep(K_MSEC(CONFIG_NET_TCP_INIT_RETRANSMISSION_TIMEOUT / 2));
```

ssharks replied:
Solved

rlubos commented on tests/net/socket/tcp/src/main.c:
```diff
@@ -709,18 +709,66 @@ void test_open_close_immediately(void)
+	/* Although closing a server socket does not require communication,
+	 * wait a little to make the test robust to scheduling order
+	 */
+	k_sleep(K_MSEC(CONFIG_NET_TCP_INIT_RETRANSMISSION_TIMEOUT/2));
```
```suggestion
	k_sleep(K_MSEC(CONFIG_NET_TCP_INIT_RETRANSMISSION_TIMEOUT / 2));
```

ssharks replied:
Solved
